content/blog/2019-11-13-space-efficient-gc/index.md

```diff
@@ -0,0 +1,100 @@
+
+### Invalid Pointer Record
+
+A record of invalid pointers found during a garbage collection that could be valid object addresses afterwards is kept as a **blacklist**. The figure below shows the modified marking algorithm with blacklisting. The only modification is in bold, which adds adjacent addresses to the blacklist.
+
+![002.png](./002.png)
+
+The garbage collector does not allocate objects at blacklisted address. This scheme would mostly blacklist addresses that correspond to long-lived data values before these values become false references as they are the data that could possibly cause garbage to be retained indefinitely. 
+
+### Implementation
+
+They implemented variants of this approach in some versions of [PCR](https://dl.acm.org/citation.cfm?id=74862) and other [garbage collectors](https://dl.acm.org/citation.cfm?id=52202). They both conservatively scan the stacks, registers, static data and the heap.
+
+Entire pages rather than individual addresses are blacklisted. In that way, the blacklist can be implemented as a bit array, indexed by page numbers. Hask table can be utilized for discontiguous heaps.
+
+### Evaluation
+
+They evaluate their techniques by runing **a** program on different machines using both statically and dynamically linked versions of C library. The program allocates 200 circular linked lists containing 100 Kbytes each. And the collector would retain the entire list if any data points to any of the 100,000 addresses corresponding to objects in the list. The results of **storage retention** with and without blacklisting are shown in the following table.
+
+![003.png](./003.png)
+
+Several observations can be made based on the table:
+
+1. Blacklisting is effective in nearly eliminating all accidental retention caused by garbage collector conservativism.
+2. The numbers in the table are approximate as the results are not completely reproducible. This is due to the fact that the scanned part of the address space is polluted with UNIX environment variables. So they are specified as ranges.
+3. If all interior pointers are considered valid, it would be difficult to allocate individual objects larger than about 100 Kbytes without violating the blacklist constraint, or requesting memory from the OS at a garbage collector specified location.
+4. According to the paper, blacklist can be easily incorpoated into a garbage collecting allocator at *almost no performance cost*.
+
+### Other Sources of Excess Retention
+
+Another source of excess memory retention is due to the fact that conservative collectors usually have less information about variable liveness than conventional collectors. As it lacks these information, the allocator could come for help.
+
+Here are some useful techniques:
+- Have the allocator and collector carefully clean up after themselves, clearing local variables before function exit.
+- The allocator try to clear areas sometimes in the stack beyond the most recently activated frame.
```

Comment:
try -> can try

Reply:
I'm not sure what "areas sometimes in the stack" means. TBH, I don't 100% understand what this part of the paper is talking about. It seems like, if you clear stack frames individually as you pop them, there's nothing else to do to the "rest of" the stack.

If you, like me, don't completely understand this part, it is totally fine to just say that here and point to exactly what's confusing.

Reply:
Also confused!!!